test: add grace period test for multi-node voting

pbeza opened the pull request:
Closes #726

DSharifi commented on crates/contract/tests/inprocess/expired_attestation.rs:
```diff
@@ -61,12 +61,21 @@ impl TestSetup {
+            .unwrap();
+    }
+
+    /// Try to submit attestation and return Result for testing failures
```
```suggestion
```

DSharifi commented on crates/contract/tests/inprocess/expired_attestation.rs:
```diff
@@ -84,11 +93,32 @@ impl TestSetup {
         }
     }
+
+    /// Get NodeIds created from the existing participants
```
```suggestion
```

DSharifi commented on crates/contract/tests/inprocess/expired_attestation.rs:
```diff
@@ -84,11 +93,32 @@ impl TestSetup {
+            .collect()
+    }
+
+    /// Helper to create attestation with hash constraints
```
```suggestion
```

DSharifi commented on crates/contract/tests/inprocess/expired_attestation.rs:
```diff
@@ -84,11 +93,32 @@ impl TestSetup {
+    }
+
+    /// Helper to create attestation with hash constraints
+    fn create_hash_attestation(hash: [u8; 32]) -> Attestation {
```
```suggestion
    fn create_attestation_with_hash_constraint(hash: [u8; 32]) -> Attestation {
```

DSharifi commented on crates/contract/tests/inprocess/expired_attestation.rs:
```diff
@@ -437,3 +467,144 @@ fn latest_image_never_expires_if_its_not_superseded() {
+        .contract
+        .allowed_code_hashes()
+        .iter()
+        .map(|h| **h)
```
Tip:
```suggestion
        .copied()
```

DSharifi commented on crates/contract/tests/inprocess/expired_attestation.rs:
```diff
@@ -437,3 +467,144 @@ fn latest_image_never_expires_if_its_not_superseded() {
+    let test_time_1 = deployment_times[2] + 3 * NANOS_IN_SECOND;
+    set_system_time(test_time_1);
+
+    let allowed_set: HashSet<_> = test_setup
```
We should `try_into` the `allowed_code_hashes` into a fixed size array, instead of `collect`ing into a `HashSet`. That way we also get a check of the ordering and uniqueness of the entries in the `assert_eq!` below.

Similar to what we have in the test above:
https://github.com/near/mpc/blob/86a84199617f4f3a8cb52dfd0c800e09f58dd3ad/crates/contract/tests/inprocess/expired_attestation.rs#L347-L354

DSharifi commented on crates/contract/tests/inprocess/expired_attestation.rs:
```diff
@@ -437,3 +467,144 @@ fn latest_image_never_expires_if_its_not_superseded() {
+    // but at T=20s it has expired and should be filtered out by allowed_code_hashes()
+    set_system_time(v1_expiry_time + NANOS_IN_SECOND); // T=20s
+
+    // Verify that only non-expired hashes remain valid at T=20s
+    // hash_v1 should have already expired, so only hash_v2 and hash_v3 should be allowed
```
```suggestion
    // T=20s
    // hash_v1 is expired. Verify that only hash_v2 and hash_v3 are allowed.
```

DSharifi commented on crates/contract/tests/inprocess/expired_attestation.rs:
```diff
@@ -437,3 +467,144 @@ fn latest_image_never_expires_if_its_not_superseded() {
+    let allowed_after_v1_expiry = test_setup.contract.allowed_code_hashes();
+    let allowed_set_after_v1_expiry: HashSet<_> =
+        allowed_after_v1_expiry.iter().map(|h| **h).collect();
+    let expected_set_after_v1_expiry: HashSet<_> = [hash_v2, hash_v3].into_iter().collect();
+
+    assert_eq!(
+        allowed_set_after_v1_expiry, expected_set_after_v1_expiry,
+        "Only hash_v2 and hash_v3 should remain valid at T=20s (hash_v1 expired)"
+    );
```
Same comment as above. Let's try_into a fixed size array.

DSharifi reviewed (approved):
Thanks @pbeza . 

Nice explaination with the ASCII art 👍

DSharifi commented:
I don't mind more tests, but don't we have very similar coverage in this existing test? It also tests that all hashes are allowed during grace periods.

https://github.com/near/mpc/blob/478b58878f788655067fae8258f87d137dde62d7/crates/contract/tests/sandbox/tee.rs#L380-L430

pbeza commented:
> I don't mind more tests, but don't we have very similar coverage in this existing test? It also tests that all hashes are allowed during grace periods.
> 
> https://github.com/near/mpc/blob/478b58878f788655067fae8258f87d137dde62d7/crates/contract/tests/sandbox/tee.rs#L380-L430

Ah, you're right. I missed that.

pbeza commented:
@DSharifi  Sorry, I merged prematurely from the GitHub mobile app, where I only saw these two review comments and no related code changes (poor UI design):
- https://github.com/near/mpc/pull/1183#pullrequestreview-3269471845
- https://github.com/near/mpc/pull/1183#issuecomment-3336042440

I’ve addressed your comments in https://github.com/near/mpc/pull/1189 instead. Feel free to review.
